# MOD102: Intro to functions

## Learning goals

1.  _Encapsulate_ logic in a function using _parameters_
2.  Improve clarity by limiting your functions' _scope_
3.  Use _immutable_ variables to reduce bugs

For this assignment, you'll submit a script that answers the questions below. Let's make sure it's a reproducible script! Your submission should run without errors on a fresh R session, which means you'll need to include appropriate function and variable definitions, as well as any necessary `library()` calls if you use a package.

-----

## What's a parameter?

### New vocabulary

- _parameter_ An input to a function 
- _encapsulate_ Take a complex procedure and hide it in a simple interface (e.g. a function)

_Parameters_ are inputs to functions. If a function is a recipe, then parameters are the ingredients. I could call `stir_fry(ingredients)` on `c("carrots", "tofu")` or on `c("chicken", "zucchini")`. You'll typically write functions when you have a piece of code you want to run multiple times on different inputs. Let's practice!

### Counting nucleotides

Congratulations, you've discovered the gene that makes koala bears so dang cute. It looks like this:

```r
koala_gene <- "AGCCTTAAATAACGACCTTC"
```

Your labmate wants to know the frequency of each nucleotide. Your challenge is to write a function that takes two parameters, `nucleo` and `sequence`, and returns the frequency of that nucleotide in the sequence. Here's a code chunk you can use to get started.

```r
nA <- gregexpr("A", koala_gene)
if (length(nA[[1]]) == 1 && nA[[1]] == -1) {
  nA <- 0
} else {
  nA <- length(nA[[1]])
}
nA
```

That code is kind of confusing for such a simple task, isn't it? It uses `gregexpr()` to search for _A_ in `koala_gene`. If you look up the help for `gregexpr()` it says:

> returns an integer vector of the same length as `text` giving the starting position of the first match or _-1_ if there is none

_(Note: that documentation is listed under the `regexpr` paragraph in the `Value` section, but it also applies to `gregexpr()` in this case)_

So first we call `gregexpr()`. Then, we check if there were no matches (i.e. we get a `-1`), in which case we set `nA` to 0. Otherwise, we set `nA` to the length of the matches vector. Woof! It'd be so much nicer if we had a function to _encapsulate_ that behavior and tuck it away for future use. Here's how we do that:

```{r}
n_nucleo <- function(nucleo, sequence) {
  result <- gregexpr(nucleo, sequence)
  if (length(result[[1]]) == 1 && result[[1]] == -1) {
    result <- 0
  } else {
    result <- length(result[[1]])
  }
  result
}
n_nucleo("A", koala_gene)
```

The code looks familiar, doesn't it? But we've changed a few key things to make the specific example more general. We _paramaterized_ the nucleotide ("A" in the example is now `nucleo`) and sequence (`koala_gene` is now `sequence`) and stuck it in a function. Calling `n_nucleo("A", koala_gene)` gets us the same result as the original. Now you have a tool for answering your labmate's question. How many of each nucleotide are there? Make a vector with four elements corresponding to A, C, G, and T frequencies, respectively.

**Q1**

```r
# Use n_nucleo()
```

### Write your own

Now your labmate wants to know the GC content of a sequence (the fraction of the sequence that's _G_ or _C_). For example, the GC content of `AAG` is 0.33 and for `TGCCCATGGG` it's 0.7. Write a function that calculates the GC content for a sequence.

**Q2**

```r
gc_content <- function(sequence) {
  
}
gc_content("AAG")
gc_content("TGCCCATGGG")
```

Now something a bit trickier. You read a paper that found the ancestral marsupial cuteness gene and you want to know how conserved it is in koalas. So you count up the number of point mutations:

```r
marsupial_gene <- "AGCCTTTCAACACGACCTTC"
koala_nucleos <- strsplit(koala_gene, "")[[1]]
marsupial_nucleos <- strsplit(marsupial_gene, "")[[1]]
n_mutations <- sum(koala_nucleos != marsupial_nucleos)
n_mutations
```

But as a wise and seasoned scientist, you know this won't be the last time you have to count point mutations. So as a service to both future you and the broader marsupial genomics community, you write a function to _encapsulate_ this logic.

**Q3**
```r
??? <- function(???) {
  ???
}
```

### Recap

Include the answers to the following questions in your script as comments.

**Q4** Choose a piece of code (~15-30 lines) and describe how you would parameterize it. This can be code you wrote, or you can search GitHub, StackOverflow, etc. In your description, specify:

* What task is the code snippet accomplishing?
* What _specific_ variables would you turn into _general_ parameters?
* What would you name this function?

**Q5** _Encapsulation_ also makes it easy to read and understand your code. Read [_Functions are for humans and computers_](https://r4ds.had.co.nz/functions.html#functions-are-for-humans-and-computers) from the R For Data Science book and take another look at your function name. Would replacing the code snippet with your function's name make the code more readable? After reading the R4DS section, would you revise the name?

_You're encouraged to take a bio break before moving on to the rest of the lesson_

-------

## What is scope?

### New vocabulary

- _scope_ Which objects (data, functions, etc) are visible to a code block 
- _immutability_ An immutable variable cannot be changed

_Scope_ is a powerful tool for breaking down complex ideas into manageable chunks, but it's also a common source of frustration for coders new to writing functions. In this lesson you'll learn what variables are visible to a function (_scope_) and how R uses _immutability_ to reduce errors.

### Bills and flippers

For this lesson we're going to examine a morphological dataset of penguins from the Palmer Station Long Term Ecological Research site in Antarctica. For more information on the R package containing this data set (and the source of the excellent penguin illustration), see https://allisonhorst.github.io/palmerpenguins/.

![Illustrations of Chinstrap, Gentoo, and AdÃ©lie penguins](https://allisonhorst.github.io/palmerpenguins/reference/figures/lter_penguins.png){width=300}

Run the following to visualize the dataset so you have a sense of what we're working with.

```r
library(tidyverse)
library(palmerpenguins)
ggplot(drop_na(penguins, flipper_length_mm, bill_length_mm), 
       aes(flipper_length_mm, bill_length_mm, color = species)) +
  geom_point(aes(shape = species), size = 2, alpha = 0.75) +
  geom_smooth(method = "lm", formula = y ~ x, se = FALSE) +
  labs(x = "Flipper length (mm)",
       y = "Bill length (mm)",
       color = "Penguin species",
       shape = "Penguin species") +
  scale_color_manual(values = c(Adelie = "#E4811E",
                                Chinstrap = "#B05CC5",
                                Gentoo = "#417175")) +
  theme_minimal() +
  theme(legend.box.background = element_rect(color = NA, fill = "white"),
        legend.justification = c(1, 0),
        legend.position = c(0.95, 0.05))
```

### A simple example of _scope_

Let's say you want the _ith_ longest bill in the dataset. You could write a function to do that. 

```r
ith_bill <- function(i, bills) {
  sort(bills, decreasing = TRUE)[i]
}
```

As you learned earlier in the lesson, `i` and `bills` are the function's parameters. You could call the function to find the 3rd longest bill.

```r
ith_bill(3, penguins$bill_length_mm)
```

**Q6**
How long is the 3rd longest bill for each of the three species?

But let's say `i` was already defined elsewhere. What would the following return, the 2nd or 10th longest?

```r
i <- 2
ith_bill(10, penguins$bill_length_mm)
```

Within the _scope_ of `ith_bill()`, `i` is 10, because that's what we passed it as a parameter. So even though we set `i` to 2 in the global environment, when we drop down into `ith_bill()`, `i` takes the value of 10.

This is another benefit of _encapsulation_. When you _encapsulate_ logic in a function, it's easier to keep track of what each variable means. You don't have to spend any mental energy figuring out "Ok, I think `i` is 10 here, but I used `i` in a for-loop elsewhere, so is it actually 10 or did it get reset in the loop?" That reduces the cognitive load and makes reading, writing, and debugging code easier.

### Pure functions

`ith_bill()` is an example of a _pure_ function. Pure functions have two properties. 

1.  The return value is always the same for a given input (e.g., no stochasticity or dependence on external data).
2.  The function has no side effects (e.g., no effect on external variables or data).

Pure functions are often desirable because these properties make them easy to reason about. Let's change `ith_bill()` to make it impure.

```r
ith_bill_impure <- function(i) {
  bills <- penguins$bill_length_mm
  sort(bills, decreasing = TRUE)[i]
}
```

We removed the `bills` parameter and instead we grab the `bill_length_mm` column from the `penguins` data frame directly. 

**Q7**

Why is `ith_bill_impure()` not a pure function? Give an example where the same input can lead to different return values.

**Q8**

Which of the following functions are pure? For the impure functions, how do they violate the properties of a pure function?

-  `sqrt()`
-  `rnorm()`
-  `lm()`
-  `read.csv()`
-  `write.csv()`

### Keeping scope limited

Most programming languages won't let you write a function like `ith_bill_impure()`, because it reaches outside of the function's scope to find the `penguins` data frame. R allows you to do that for convenience, but it can lead to some really gnarly, difficult to diagnose bugs. You're definitely going to have to write impure functions sometimes, like reading a CSV file. But just because R will let you write impure functions doesn't mean you have to! 

**Q9**

The following code clusters penguins by morphology and assesses what fraction are misclassified by species. Re-write it in three pure functions: `prepare_data()`, `find_clusters()`, and `assess_clusters()`. If you don't understand how the code works, ask another student or TA for assistance.

```r
penguin_morpho <- select(penguins, bill_length_mm, flipper_length_mm, species)
penguin_morpho <- drop_na(penguin_morpho)
penguin_clust <- kmeans(select(penguin_morpho, -species), 
                        centers = n_distinct(penguins$species))
penguin_morpho$cluster <- penguin_clust$cluster
cluster_species <- penguin_morpho %>% 
  group_by(species) %>% 
  summarize(cluster = names(which.max(table(cluster)))) %>% 
  transmute(clustered_species = species, 
            cluster = as.numeric(cluster))
penguin_morpho <- left_join(penguin_morpho, cluster_species, by = "cluster")
n_misclass <- sum(penguin_morpho$species != penguin_morpho$clustered_species)
n_total <- nrow(penguin_morpho)
n_misclass / n_total
```

In the script for Q9, `penguin_morpho` was altered multiple times. This adds extrinsic cognitive load because to understand `penguin_morpho` you have to hold multiple operations in your memory at once. It's also a really common source of bugs! In the last part of this lesson, you'll learn how to treat variables as _immutable_, which will probably feel counter-intuitive but in the long run it makes code easier to read, write, and share. But first, take a bio-break. 

-----

## Leave that variable alone

A mutable variable can be changed, an _immutable_ variable cannot. Not being able to change a variable seems like a major limitation, so why would immutable variables be preferred? From the [MIT course](https://web.mit.edu/6.005/www/fa15/classes/09-immutability/) on software construction:

> The answer is that **immutable types are safer from bugs, easier to understand, and more ready for change**. Mutability makes it harder to understand what your program is doing, and much harder to enforce contracts.

R objects are (mostly) immutable by default. What does "mostly" mean here? Consider the following:

```r
x <- list(1, 2, 4)
y <- x
x[[2]] <- 8
```

What's `y[[2]]`? If `y` was mutable, it would be 8. But lists are (mostly) immutable, so when we modified `x`, `y` retained its original values. This makes it much easier to reason about your code, because assignment affects exactly one variable. When we run `x[[2]] <- 8`, we don't have to worry if there are any side effects, like other variables changing in value.

You're also familiar with immutability from the dplyr package. Functions like `select()`, `mutate()`, and `filter()` treat their inputs as immutable. In the following code, `filter()` returns a subset of the `penguins` data frame, but the original data frame is left untouched.

```r
filter(penguins, species == "Gentoo")
penguins
```

There are a lot of other benefits to immutability. Consider the pipe operator (`%>%`). This is one of the most useful tools for making data munging code easier to read. If I need to run a series of operations on a data frame, I can pipe the operations together without worrying if I'm altering the original data.

```r
# What's the average bill size of penguins with longer-than-average flippers?
# (by species)
penguins %>% 
  group_by(species) %>% 
  filter(flipper_length_mm > mean(flipper_length_mm, na.rm = TRUE)) %>% 
  summarize(mean_bill_length = mean(bill_length_mm, na.rm = TRUE))
```

Like _scope_, using _immutability_ to our advantage requires us to constrain ourselves. We _can_ look outside a function's scope for variables, just as we _can_ alter variables as we  go. But if we constrain ourselves to follow the rules of pure functions (stay in your lane, no side effects) as often as possible, we'll write easy-to-read code with fewer bugs.

**Q10**

Fill out the parameters and body of `body_condition()` and `condition_trend()` in the following code. Make sure your functions are pure, i.e., limited scope and immutable variables. If you've written the functions correctly, you should get a plot like this:

![Plot result for Q10](Q10.png)

```r
# body_condition() quantifies the body condition of a penguin as its residual
# (normalized) mass relative to its structural size (bill length, bill depth, 
# and flipper length). Use lm() to fit a regression and resid() to get the 
# residuals. 
body_condition <- function(???, ???, ???, ???) {
  mass_normalized <- (??? - mean(???)) / sd(???)
  mod <- lm(??? ~ ??? + ??? + ???)
  resid(mod)
}

penguins_condition <- penguins %>% 
  drop_na() %>% 
  group_by(species) %>% 
  mutate(condition = body_condition(???, ???, ???, ???)) %>% 
  ungroup()

# condition_trend() calculates the slope of body condition over time. Use lm() 
# to fit a regression and coef() to get the slope.
condition_trend <- function(???, ???) {
  mod <- lm(??? ~ ???)
  coef(mod)[???]
}

penguins_trend <- penguins_condition %>% 
  group_by(species) %>% 
  summarize(trend = condition_trend(???, ???)) %>% 
  mutate(trend_lbl = sprintf("slope=%0.2f", trend))
  
ggplot(penguins_condition, aes(factor(year), condition)) + 
  geom_violin() + 
  geom_smooth(aes(group = species), method = "lm", se = FALSE) + 
  geom_text(aes(label = trend_lbl), penguins_trend, x = 2, y = 0.05) +
  facet_grid(rows = vars(species)) +
  theme_minimal()
```

## Submit assignment

[Create a gist](https://docs.github.com/en/github/writing-on-github/editing-and-sharing-content-with-gists/creating-gists) from the script you wrote for this lesson. Open an issue in [FlukeAndFeather/jese4sci-MOD](https://github.com/FlukeAndFeather/jese4sci-MOD). Put "MOD102" in the title. In the comment, provide a link to your gist.
